src/luascript.cpp

```diff
@@ -9391,55 +9391,52 @@ int LuaScriptInterface::luaPlayerHasMount(lua_State* L) {
 	return 1;
 }
 
-int LuaScriptInterface::luaPlayerGetPremiumDays(lua_State* L)
+int LuaScriptInterface::luaPlayerGetPremiumTime(lua_State* L)
 {
-	// player:getPremiumDays()
+	// player:getPremiumTime()
 	Player* player = getUserdata<Player>(L, 1);
 	if (player) {
-		lua_pushnumber(L, player->premiumDays);
+		lua_pushnumber(L, std::max<int32_t>(0, player->premiumEndsAt - time(nullptr)));
 	} else {
 		lua_pushnil(L);
 	}
 	return 1;
 }
 
-int LuaScriptInterface::luaPlayerAddPremiumDays(lua_State* L)
+int LuaScriptInterface::luaPlayerAddPremiumTime(lua_State* L)
 {
-	// player:addPremiumDays(days)
+	// player:addPremiumTime(seconds)
 	Player* player = getUserdata<Player>(L, 1);
 	if (!player) {
 		lua_pushnil(L);
 		return 1;
 	}
 
-	if (player->premiumDays != std::numeric_limits<uint16_t>::max()) {
-		uint16_t days = getNumber<uint16_t>(L, 2);
-		int32_t addDays = std::min<int32_t>(0xFFFE - player->premiumDays, days);
-		if (addDays > 0) {
-			player->setPremiumDays(player->premiumDays + addDays);
-			IOLoginData::addPremiumDays(player->getAccount(), addDays);
-		}
-	}
+	uint32_t seconds = getNumber<uint32_t>(L, 2);
+	int32_t oldTime = std::max<int32_t>(0, player->premiumEndsAt - time(nullptr));
+	player->setPremiumTime(time(nullptr) + oldTime + seconds);
+	IOLoginData::updatePremiumTime(player->getAccount(), player->premiumEndsAt);
 	pushBoolean(L, true);
 	return 1;
 }
 
-int LuaScriptInterface::luaPlayerRemovePremiumDays(lua_State* L)
+int LuaScriptInterface::luaPlayerRemovePremiumTime(lua_State* L)
 {
-	// player:removePremiumDays(days)
+	// player:removePremiumTime(seconds)
 	Player* player = getUserdata<Player>(L, 1);
 	if (!player) {
 		lua_pushnil(L);
 		return 1;
 	}
 
-	if (player->premiumDays != std::numeric_limits<uint16_t>::max()) {
-		uint16_t days = getNumber<uint16_t>(L, 2);
-		int32_t removeDays = std::min<int32_t>(player->premiumDays, days);
-		if (removeDays > 0) {
-			player->setPremiumDays(player->premiumDays - removeDays);
-			IOLoginData::removePremiumDays(player->getAccount(), removeDays);
-		}
+	uint32_t seconds = getNumber<uint32_t>(L, 2);
+	int32_t oldTime = std::max<int32_t>(0, player->premiumEndsAt - time(nullptr));
+	if (oldTime < seconds) {
+		player->setPremiumTime(0);
+		IOLoginData::updatePremiumTime(player->getAccount(), 0);
+	} else {
+		player->setPremiumTime(time(nullptr) + (oldTime - seconds));
+		IOLoginData::updatePremiumTime(player->getAccount(), player->premiumEndsAt);
 	}
```

Comment:
I think what you're doing here is that you're resetting the `premiumEndsAt` timestamp to 0, whenever the premium time is set to a time in the past. That does not sound like expected behavior.

Imagine a Lua script that removes 7 days of premium and then adds 7 days of premium for someone who has 6 days of premium left. With your logic, their premium would now expire 6 days after epoch (in the past) instead of 6 days from now.

```suggestion
	if (player->premiumEndsAt < seconds) {
		player->setPremiumTime(0);
		IOLoginData::updatePremiumTime(player->getAccount(), 0);
	} else {
		player->setPremiumTime(player->premiumEndsAt - seconds);
		IOLoginData::updatePremiumTime(player->getAccount(), player->premiumEndsAt);
	}
```